File: week-02/yc_3015_oasis.cs
using System;
using System.Collections.Generic;

record Person(int Height, long Count);

class Program
{
    static void Main(String[] args)
    {
        int n = int.Parse(Console.ReadLine());
        Stack<Person> stack = new Stack<Person>();
        long result = 0;

        for (int i = 0; i < n; i++)
        {
            int h = int.Parse(Console.ReadLine());
            long count = 1;

            while (stack.Count && stack.Peek().Height < h)
            {
                result += stack.Peek().Count;
                stack.Pop();
            }

            if (stack.Count > 0 && stack.Peek().Height == h)
            {
                var top = stack.Pop();
                result += top.Count;
                count = top.Count + 1;

                if (stack.Count > 0)
                {
                    result += 1;
                }
            }
            else
            {
                if (stack.Count > 0)
                {
                    result += 1;
                }
            }
            stack.Push(new Person(h, count));
        }
        Console.WriteLine(result);
    }
}

File: week-02/yc_3015_oasis.java
import java.io.*;
import java.util.*;

public class Main {
    record Person(int height, long count) {}

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        Stack<Person> stack = new Stack<>();
        long result = 0;

        for (int i = 0; i < N; i++) {
            int h = Integer.parseInt(br.readLine());
            long count = 1;

            while (!stack.isEmpty() && stack.peek().height() < h) {
                result += stack.peek().count();
                stack.pop();
            }

            if (!stack.isEmpty() && stack.peek().height() == h) {
                Person top = stack.pop();
                result += top.count();
                count = top.count() + 1;

                if (!stack.isEmpty()) {
                    result += 1;
                }
            } else {
                if (!stack.isEmpty()) {
                    result += 1;
                }
            }

            stack.push(new Person(h, count));
        }

        System.out.println(result);
    }
}
